tools/java/org/unicode/cldr/util/XPathParts.java

```diff
@@ -439,7 +440,19 @@ public String getAttributeValue(int elementIndex, String attribute) {
     }
 
     public void putAttributeValue(int elementIndex, String attribute, String value) {
-        elements.get(elementIndex >= 0 ? elementIndex : elementIndex + size()).putAttribute(attribute, value);
+        elementIndex = elementIndex >= 0 ? elementIndex : elementIndex + size();
+        if (value == null && (elements.get(elementIndex).attributes == null ||
+                              !elements.get(elementIndex).attributes.containsKey(attribute))) {
+            return;
+        }
+        if (value != null && elements.get(elementIndex).attributes != null
+            && elements.get(elementIndex).attributes.containsKey(attribute)
+            && elements.get(elementIndex).attributes.get(attribute).equals(value)) {
+            return;
+        }
+        makeElementsMutable();
+        makeElementMutable(elementIndex);
+        elements.get(elementIndex).putAttribute(attribute, value);
```

Comment:
Add
// make mutable may change elements.get(elementIndex), so we have to use elements.get(elementIndex) after calling.